Improve Google Closure Library Bundling

pyrmont opened the pull request:
Planck bundles the Google Closure Library. Previous to this PR, that was achieved by: (1) installing an unofficial distribution of GCL from Maven Central; and (2) manually listing the namespaces to import in `planck-cljs/bundle.cljs`. This was problematic since it meant that the GCL bundled with Planck was often not up to date. It also was difficult to maintain as the large number of namespaces made available via GCL were hand-coded.

This PR solves those problems by having the build script: (1) download an official distribution of GCL from the official GitHub repository; and (2) generate the namespaces to import from the files
that make up this distribution.

In addition, this PR also introduces a caching mechanism for the files generated as part of the build process. Planck provides a `--fast` compilation mode that skips some of the optimisation steps that greatly increase build times but even this mode repeats work that is unnecessary for the majority of the bundled files that do not change.

The cache works by saving the generated files to the cache with a name that combines the name of the file and the MD5 hash of the file's contents.

pyrmont commented:
Outstanding issues:

- Although it formed the impetus for developing this branch, the PR doesn't actually provide a mechanism to download a pre-optimised cache. You can [download a version I generated] to see how it works.

- The build process stalled several times during development. I couldn't determine the source of the problem. I'm doing my development on a Linode instance and it may have been related to resource sharing amongst other instances but I'm not sure.

- The CI builds fail. I think this is because of the way that the GCL is bundled in (and possibly also the use of the build cache). I haven't looked into this yet.

mfikes commented:
It seems that, with the way things are set up, we could put a pre-built cache on planck-repl.org and have the build script attempt to fetch it and put it in place if it doesn't already exist locally. (To prime the cache, so to speak.)

Then, during dev, as Planck source evolves, small parts of that pre-built cache would just be invalidated with fresh cache files being created.

If we did this, I would expect this to work with Ubuntu PPA builds (which don't allow network access during the build) because I would cause things to prime the cache when going through the steps at https://github.com/planck-repl/ubuntu-ppa by removing `FAST_BUILD=1` from step 13 and thus the primed cache would be included in the archive produced in step.

If you wanted to just wrap up this PR and not deal with this stuff in this PR, I could take this on later. Up to you.

I think this would be nice because it would make things fast for people who happen to clone planck and build it, or for people who end up doing `brew install --HEAD planck`.

pyrmont commented:
A few more notes:

- I've fixed the CI build issue. The GCL namespaces are defined in a new file in `planck-cljs/src/planck/bundle/gcl.cljs` but I wasn't creating the `bundle` directory and so if that didn't already exist, the build would fail.

- The initial commit changed the parallel build flag to `false`. This was an attempt to solve the problem I mentioned earlier where the optimisation stage would sometimes freeze. After thinking more about it, I don't think this could be related to that issue (and the freezes kept happening anyway) so I have changed this flag back to `true`.

- I've updated the GCL documentation to refer to the correct namespaces that are excluded. I've also added a note explaining how a user can alter these for custom builds of Planck.

- I've fixed some bugs in the `clean` and `bundle` Bash scripts.

In terms of the distribution a pre-built cache, I'd like to do that but think it would be best to implement it through a separate PR. If you can tell me the URL for the archive file once you've uploaded it, I'll get to work on a branch that brings this in (and references #893).

mfikes commented:
Being merged in via #901
